### PR TITLE
XD-3270 - Implementation of ModuleResolver

### PR DESCRIPTION
**Pull request by dturanski**

### CODE REVIEW

File: spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+		LocalRepository localRepo = new LocalRepository(localRepoPath);
+		session.setLocalRepositoryManager(system.newLocalRepositoryManager(session, localRepo));
+		return session;
+	}
+
+	private RepositorySystem newRepositorySystem() {
+		/*
+		 * Aether's components implement org.eclipse.aether.spi.locator.Service to ease manual wiring and using the
+         * prepopulated DefaultServiceLocator, we only need to register the repository connector and transporter
```
Comment (markfisher) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java]: funky indentation here

**Comment by dturanski** [comment]
@dsyer - not sure what's going on with travis.  Builds fine locally.

File: spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * An implementation of ModuleResolver using <a href="http://www.eclipse.org/aether/>aether</a> resolve the module
+ * artifact (uber jar) in a local Maven repository, downloading the latest update from a remote repository if
+ * necessary.
+ * @author David Turanski
+ */
+public class DefaultModuleResolver implements ModuleResolver {
```
Comment (markfisher) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java]: Maybe this should be named AetherModuleResolver?... can still be our default :)

File: spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+					"runtime"));
+		}
+		catch (ArtifactResolutionException e) {
+			throw new RuntimeException(e);
+		}
+
+		return new FileSystemResource(result.getArtifact().getFile());
+	}
+
+	public DefaultModuleResolver(File localRepo, Map<String, String> remoteRepos) {
```
Comment (markfisher) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java]: should move constructor to top

File: spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+	private static final String DEFAULT_EXTENSION = "jar";
+
+	private final File localRepository;
+
+	private final List<RemoteRepository> remoteRepositories;
+
+	private final RepositorySystem repositorySystem;
+
+	@Override
+	public Resource resolve(String groupId, String artifactId, String version) {
```
Comment (markfisher) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java]: javadoc

File: spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java (hunk 3 above)
Comment (markfisher) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java]: javadoc also

File: spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java (hunk 2 above)
Comment (dturanski) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java]: I thought about that. Happy to change the name

File: spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+		RepositorySystemSession session = newRepositorySystemSession(repositorySystem,
+				localRepository.getAbsolutePath());
+		ArtifactResult result;
+
+		try {
+			result = repositorySystem.resolveArtifact(session, new ArtifactRequest(artifact, remoteRepositories,
+					"runtime"));
+		}
+		catch (ArtifactResolutionException e) {
+			throw new RuntimeException(e);
```
Comment (markfisher) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java]: Perhaps we should introduce a `ModuleResolutionException`. We could pass the coordinates to the constructor and display those in the message (and still have the wrapped exception as the root cause), although in this case - which might be our only implementation - the wrapped Exception provides that info already. Another alternative would be `IllegalArgumentException` along with corresponding JavaDoc explaining that might happen if the coordinates are unresolvable. That said, I don't view this as an *illegal argument* per se, just one that is not resolvable.

Comment (markfisher) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/DefaultModuleResolver.java]: I think we can leave it as-is for now.

File: spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/AetherModuleResolver.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+		repositorySystem = newRepositorySystem();
+	}
+
+	/**
+	 * Resolve an artifact and return its location in the local repository. Aether performs the normal
+	 * Maven resolution process ensuring that the latest update is cached to the local repository.
+	 * @param groupId the groupId
+	 * @param artifactId the artifactId
+	 * @param version the version
+	 * @return a {@ link FileSystemResource} representing the resolved artifact in the local repository.
```
Comment (markfisher) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/AetherModuleResolver.java]: should add a @throws here as well?

Comment (dturanski) [inline spring-cloud-stream-module-launcher/src/main/java/org/springframework/cloud/stream/module/resolver/AetherModuleResolver.java]: done

**Comment by markfisher** [comment]
squashed and merged as https://github.com/spring-cloud/spring-cloud-stream/commit/262d82387e3e792f0f51242b27de86165a5670d5

**Pull request closed without merging**